### DISCUSSION TITLE
Episode 122: ChatGPT, GTC 2023 & Stickers on our Foreheads

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#13] [Episode]
Episode Link: https://adspthepodcast.com/2023/03/24/Episode-122.html

Feel free to write comments or ask questions 🙂

**Comment by TylerHillery** [comment]
> **All that stuff can be a couple stickers on my forehead and a chip in my wrist.** 

This is up there as one of my favorite ADSP quotes

**Comment by jdebache** [comment]
Totally agree with the comments on using ChatGPT to help with build engineering. I use it to write Bazel rules. It's not perfect, but if I had to do it myself and dig through the doc it would be simply prohibitive.

In general, it's a great tool to deal with things that work with a lot of what I would call 'magic spells' like CLIs and build tools. A lot of things that I had given up on using in the past are now in my reach, which is really cool and exciting.

**Comment by tothambrus11** [comment]
Hello Bryce,

I'm Ambrus Tóth from Hungary, a first-year CS student in the Netherlands. I've been following your podcast since October, and now I'm finally caught up with it. It's been a fascinating journey; I really enjoyed all the discussions, arguments, and various perspectives, along with the occasional chaos, of course. Particularly, it's been quite enlightening for me to peek behind the scenes of the C++ committee, contemplating potential improvements, identifying limiting factors, and thinking about possible solutions.

I'm writing to you now because I've been deeply considering the idea of writing GitLab issues and seeing AI create PRs for them ever since you mentioned it. Initially, I presented a detailed plan to my team at Huawei on a Saturday afternoon as an April Fool's joke. However, as I strived to come up with convincing details, I began to realize that this could indeed be a viable idea, even with our current technology.

Both Copilot and GPT-4 have significantly boosted my development productivity. I've been learning JavaFX and implemented a relatively complex UI for a university project with the assistance of these tools. The process of copying, pasting, and recompiling code, pasting compiler errors, runtime exceptions, code from related classes, and other possibly related information took a considerable amount of time.

The following quarter, we took a course on software quality and testing, where we learned about a vast range of concepts and topics that can help us ensure the quality of our code and instill confidence that our code will function properly. Now, isn't this the same challenge that people using Copilot face? Developers, whether they are human or AI, aren't infallible.

Developers often struggle to know whether the AI-produced code is correct, just appears correct (for instance, they wouldn't spot the error by just looking at a PR), or appears incorrect but the developers are too casual to scrutinize it carefully.

Here's where it gets exciting... If the AI could adhere to all the software engineering methods taught in my course, it could likely derive a correct and functional solution by itself, regardless of the model's potential limitations.

By applying a modified version of test-driven development, it could:

- Design the interface of a method based on the requirements [optional]
- Apply TDD cycles:
    - Write failing TDD-like test cases for interesting cases based on the specification
    - Develop the code until the test case succeeds
        - Run the code and review compile errors and test failure messages
    - Refactor the method for enhanced readability
- Apply boundary testing based on the specification and then based on the code
- Let another agent and then a person review the spec tests, reducing the chance that the tests misrepresent the actual specification

All of these could be integrated into our favorite IDEs (and perhaps CI platforms), to minimize the manual copy-pasting and explaining.

There are many caveats here, like how will it know what functions are available for calling inside a method, how we can provide just the right context, and so on. I haven't delved deeply into this yet, but I believe this could be addressed by using the language server and potentially employing multiple sub-agents to summarize file contents based on a specific query from the main agent. The goal is to provide the same kind of lookup functionality that we humans employ during development.

I noticed that you've given talks at conferences about AI-assisted software engineering, but I couldn't find any published videos or presentations about them. I would be very interested in hearing your thoughts on this topic, whether through a recording of your talk, a future ADSP episode, or just in text format. I'm excited to start working on this project this summer with my friend, Markas Aišparas, after conducting further research and brainstorming more ideas.

**Reply by codereport** [reply]
@brycelelbach
